### ISSUE TITLE
OpenAPI (swagger) integration

### ISSUE THREAD

**Issue by szuecs** [zalando/skipper#418]
As **application developer**, I want to point to my OpenAPI endpoint and get many features for free, that I already defined in my OpenAPI spec.

Features that could be implemented with skipper as a sidecar and a defined a known skipper known OpenAPI spec endpoint:
- [ ] automated OAuth2/JWT authnz
- [ ] whitelist available routes
- [ ] request validation (regexps for http json body parameters, querystring, ..)
- [ ] .... *add features that you can think of here*

## config

required input
* flag/ENV: --open-api to enable OpenAPI sidecar feature 
* POD annotation: 
  * key: "zalando.org/openapi"
  * value: URL to find the OpenAPI spec, for example the default could be "http://localhost/.well-known/schema-discovery"

goroutine
1. register SIGTERM handler, that: disables health check endpoint and just wait for SIGKILL
1. get my POD spec to get optional "zalando.org/openapi" annotation
1. get OpenAPI spec on initial start: poll as long for example network or the app container is not yet responding
1. compile spec to routes with filters to apply
1. enable health check endpoint/route
1. end goroutine

## example flow
What happens on incoming calls is highly dependent on the OpenAPI spec, for example a `GET /foo/:bar` with authnz ("security" oauth "scope" and "securityDefinitions")?
1. If GET /foo is not present, then return 404, else continue
1. apply the filters provided by the route, that will
  1. authnz based on the Bearer token (OAuth/JWT)
  1. validate parameter bar is valid with defined regexps from the spec
  1. strip token header
  1. proxy request to target application
  1. return result of the target application to the caller

## Questions
- [ ] What OAuth flows makes sense to implement (depends on #413 and #366)?
- [ ] Check that http error codes in the example make sense

**Comment by szuecs** [comment]
related #239

**Comment by herojan** [comment]
@szuecs Is this issue still planned? We have considered making a gateway in front of our own apps that would handle auth, rate-limiting based on the token's uid, metrics, etc but it's very similar to what's planned here so if skipper could do it it would be great.

**Comment by szuecs** [comment]
@herojan yes, but time is limited so I can't give you any ETA for the complete implimentation specified in this issue. 
Can you as user answer the questions, that I specified in the issue? It would be great to understand more the user side. 
OAuth2 tokeninfo is already impleemnted and documented, if you need tokenintrospection (rfc7662) or OIDC UI flows, then I have a current PR, which I slowly work on (some other tasks have higher priorities).
As you are a Zalando user, as far as I remember there is one internal swagger-skipper implementation from some team. Please ask in our internal chat to get more information about that.

**Comment by herojan** [comment]
@szuecs Didn't realise which github this was :) 

We're currently looking at k8s service to service calls, where obtaining auth tokens for sending requests is already solved for us via secrets. So our particular use case is to bring jwt validation and rate-limiting based on a whitelist of uids, to outside of each app and not have to solve it for each new app.

**Comment by szuecs** [comment]
@herojan jwt validation can be already done via oauthTokenInfo filters see the docs for the filters: https://opensource.zalando.com/skipper/filters/#oauthtokeninfoanyscope
how to use filters from ingress you can see at https://opensource.zalando.com/skipper/kubernetes/ingress-usage/#filters-and-predicates
For ratelimits (per skipper instance until we have done #593 please tell us if this is a feature you like to have, such that we increase priority), check: https://opensource.zalando.com/skipper/kubernetes/ingress-usage/#ratelimits

**Comment by szuecs** [comment]
I think we should not implement this in skipper but there can be controllers doing the work and create ingress, crd or plain skipper eskip routes. Internally we already have something like that.
